hydrolib/hydromt_delft3dfm/workflows/boundaries.py

```diff
@@ -0,0 +1,272 @@
+# -*- coding: utf-8 -*-
+
+import configparser
+import logging
+
+import geopandas as gpd
+import hydromt.io
+import numpy as np
+import pandas as pd
+import xarray as xr
+import shapely
+from hydromt import config
+from scipy.spatial import distance
+from shapely.geometry import LineString, Point
+
+from .graphs import gpd_to_digraph
+
+logger = logging.getLogger(__name__)
+
+
+__all__ = [
+    "generate_boundaries_from_branches",
+    "select_boundary_type",
+    "validate_boundaries",
+    "compute_boundary_values",
+]
+
+
+def generate_boundaries_from_branches(branches: gpd.GeoDataFrame, where: str = "both"):
+    """function to get possible boundary locations from branches with id
+
+    [ ] convert branches to graph
+    [ ] get boundary locations at where
+
+    parameters:
+    where: str
+        Options available aare: ['upstream', 'downstream', 'both']
+        Where at the branches should the boundaries be derived.
+    """
+
+    # convert branches to graph
+    G = gpd_to_digraph(branches)
+
+    # get boundary locations at where
+    if where == "downstream":
+        endnodes = {
+            dn: {**d, **{"where": "downstream"}}
+            for up, dn, d in G.edges(data=True)
+            if G.out_degree[dn] == 0 and G.degree[dn] == 1
+        }
+    elif where == "upstream":
+        endnodes = {
+            up: {**d, **{"where": "upstream"}}
+            for up, dn, d in G.edges(data=True)
+            if G.in_degree[up] == 0 and G.degree[up] == 1
+        }
+    elif where == "both":
+        endnodes = {
+            dn: {**d, **{"where": "downstream"}}
+            for up, dn, d in G.edges(data=True)
+            if G.out_degree[dn] == 0 and G.degree[dn] == 1
+        }
+        endnodes.update(
+            {
+                up: {**d, **{"where": "upstream"}}
+                for up, dn, d in G.edges(data=True)
+                if G.in_degree[up] == 0 and G.degree[up] == 1
+            }
+        )
+    else:
+        pass
+
```

Comment:
Or:

```suggestion
    endnodes = {}
    if where == "both" or where == "downstream":
        endnodes.update(
            {
                dn: {**d, **{"where": "downstream"}}
                for up, dn, d in G.edges(data=True)
                if G.out_degree[dn] == 0 and G.degree[dn] == 1
            }
        )

    if where == "both" or where == "upstream":
        endnodes.update(
            {
                up: {**d, **{"where": "upstream"}}
                for up, dn, d in G.edges(data=True)
                if G.in_degree[up] == 0 and G.degree[up] == 1
            }
        )
```

to remove duplicity and complecxity